### PR TITLE
NET-291 PUT /api/networks/{networkname}

### PR DESCRIPTION
**Pull request by TobiaszCudnik**
## Describe your changes

NET-291 Added `PUT /api/networks/{networkname}`

## Provide testing steps

https://toil.kitemaker.co/cQARgD-Gravitl/6a50SD-Netmaker/items/291

## Checklist before requesting a review
- [ ] My changes affect only 10 files or less.
- [ ] I have performed a self-review of my code and tested it.
- [ ] If it is a new feature, I have added thorough tests, my code is <= 1450 lines.
- [ ] If it is a bugfix, my code is <= 200 lines.
- [ ] My functions are <= 80 lines.
- [ ] I have had my code reviewed by a peer.
- [ ] My unit tests pass locally.
- [ ] Netmaker is awesome.

### CODE REVIEW

File: controllers/network.go [hunk 1]
```diff
@@ -282,3 +283,57 @@ func createNetwork(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusOK)
 	json.NewEncoder(w).Encode(network)
 }
+
+// swagger:route PUT /api/networks networks updateNetwork
+//
+// Create a network.
```
Comment (Aceix) [inline controllers/network.go:289]: updates a network

File: controllers/network.go [hunk 2]
```diff
@@ -282,3 +283,57 @@ func createNetwork(w http.ResponseWriter, r *http.Request) {
...
+func updateNetwork(w http.ResponseWriter, r *http.Request) {
+
+	w.Header().Set("Content-Type", "application/json")
+
+	var netNew models.Network
+
+	// we decode our body request params
+	err := json.NewDecoder(r.Body).Decode(&netNew)
+	if err != nil {
+		logger.Log(0, r.Header.Get("user"), "error decoding request body: ",
```
Comment (Aceix) [inline controllers/network.go:307]: use slog

File: controllers/network.go [hunk 3]
```diff
@@ -282,3 +283,57 @@ func createNetwork(w http.ResponseWriter, r *http.Request) {
...
+			err.Error())
+		logic.ReturnErrorResponse(w, r, logic.FormatError(err, "badrequest"))
+		return
+	}
+
+	if netNew.AddressRange == "" && netNew.AddressRange6 == "" {
+		err := errors.New("IPv4 or IPv6 CIDR required")
+		logger.Log(0, r.Header.Get("user"), "failed to create netNew: ",
```
Comment (Aceix) [inline controllers/network.go:315]: slog

File: controllers/network.go [hunk 4]
```diff
@@ -282,3 +283,57 @@ func createNetwork(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	netOld, err := logic.GetNetwork(netNew.NetID)
+	if err != nil {
+		logger.Log(0, r.Header.Get("user"), "error fetching network: ",
+			err.Error())
+		logic.ReturnErrorResponse(w, r, logic.FormatError(err, "badrequest"))
+		return
+	}
+	_, _, _, _, _, err = logic.UpdateNetwork(&netOld, &netNew)
```
Comment (Aceix) [inline controllers/network.go:328]: can the update be restricted to only groups and users

**Review by Aceix — CHANGES_REQUESTED** [review]

**Review by mattkasun — CHANGES_REQUESTED** [review]

File: controllers/network.go [hunk 5]
```diff
@@ -282,3 +283,57 @@ func createNetwork(w http.ResponseWriter, r *http.Request) {
...
+	// we decode our body request params
+	err := json.NewDecoder(r.Body).Decode(&netNew)
+	if err != nil {
+		logger.Log(0, r.Header.Get("user"), "error decoding request body: ",
+			err.Error())
+		logic.ReturnErrorResponse(w, r, logic.FormatError(err, "badrequest"))
+		return
+	}
+
+	if netNew.AddressRange == "" && netNew.AddressRange6 == "" {
```
[the remark this answers is not in the document]
Comment (TobiaszCudnik) [reply controllers/network.go]: Limited to `ProSettings` only.

File: controllers/network.go (hunk 1 above)
Comment (TobiaszCudnik) [reply controllers/network.go]: Fixed.

File: controllers/network.go (hunk 2 above)
Comment (TobiaszCudnik) [reply controllers/network.go]: Fixed.

File: controllers/network.go (hunk 4 above)
Comment (TobiaszCudnik) [reply controllers/network.go]: Limited to `ProSettings` now.

**Pull request merged**